contracts/arbitrage/BancorArbitrage.sol

```diff
@@ -486,4 +474,18 @@ contract BancorArbitrage is ReentrancyGuardUpgradeable, Utils, Upgradeable {
 
         emit ArbitrageExecuted(caller, exchangeIds, path, sourceAmount, burnAmount, rewardAmount);
     }
-}
+
+    /**
+     * @dev set exchange allowance to the max amount if it's less than the input amount
+     */
+    function _setExchangeAllowance(Token token, address exchange, uint inputAmount) private {
+        if(token.isNative()) {
+            return;
+        }
+        uint allowance = token.toIERC20().allowance(address(this), exchange);
+        if(allowance < inputAmount) {
+            // increase allowance to the max amount if allowance < inputAmount
+            token.safeIncreaseAllowance(exchange, type(uint256).max - allowance);
```

Comment:
First off, you can probably use `token.safeApprove(exchange, type(uint256).max)` here.

This would save the gas used inside `safeIncreaseAllowance` in order to read the current allowance.

Second, why not use `safeIncreaseAllowance(exchange, inputAmount - allowance)`?

Or following the first comment - even `token.safeApprove(exchange, inputAmount)`?

The first reason is that the user will get more gas back after `transferFrom` is executed, because it will reset the allowance to zero (rather than just reducing it to some non-zero value).

But the more important reason is that an unlimited allowance is a potential future problem, for example, a hack in this contract would "grant" the unlimited allowance to the hacker.

Reply:
The security issue is with regards to what we don't know, not with regards to what we know.
Any hack (which we are not aware of), could potentially enjoy the unlimited allowance.

For example:
```
_setExchangeAllowance(Token(tokenIn), address(_uniswapV3Router), sourceAmount);
```
How do we ensure that an upgraded version of `_uniswapV3Router` doesn't transfer the entire allowance out of our contract?

Also, if we do decide to go ahead with this approach, then why not give unlimited approval to all known recipients, immediate upon the construction (deployment) of this contract?
Then we don't need to do any more calls to read the allowance, let alone to write it.